libs/isograph-react/src/core/optimisticProxy.ts

```diff
@@ -0,0 +1,386 @@
+      node.startUpdate = () => {
+        prevStartUpdate(node);
+        startUpdate(node);
+      };
+
+      startUpdate(node);
+      return node;
+    }
+    case 'NetworkResponseStoreLayer':
+    case 'OptimisticStoreLayer': {
+      const node: StartUpdateStoreLayer = {
+        kind: 'StartUpdateStoreLayer',
+        parentStoreLayer: parent,
+        childStoreLayer: null,
+        data: {},
+        startUpdate: startUpdate,
+      };
+      parent.childStoreLayer = node;
+
+      startUpdate(node);
+      return node;
+    }
+    default: {
+      parent satisfies never;
+      throw new Error('Unreachable. This is a bug in Isograph.');
+    }
+  }
+}
+
+export function addOptimisticStoreLayer(
+  parent: StoreLayer,
+  startUpdate: DataUpdate,
+) {
+  switch (parent.kind) {
+    case 'BaseStoreLayer':
+    case 'StartUpdateStoreLayer':
+    case 'NetworkResponseStoreLayer':
+    case 'OptimisticStoreLayer': {
+      const data = {};
+
+      const node: OptimisticStoreLayer = {
+        kind: 'OptimisticStoreLayer',
+        parentStoreLayer: parent,
+        childStoreLayer: null,
+        data,
+        startUpdate: startUpdate,
+      };
+
+      startUpdate(node);
+      parent.childStoreLayer = node;
+
+      return {
+        node,
+        revert: (
+          environment: IsographEnvironment,
+          normalizeData: (storeLayer: StoreLayer) => EncounteredIds,
+        ): void => {
+          replaceOptimisticStoreLayerWithNetworkResponseStoreLayer(
+            environment,
+            node,
+            normalizeData,
+          );
+        },
+      };
+    }
+    default: {
+      parent satisfies never;
+      throw new Error('Unreachable. This is a bug in Isograph.');
+    }
+  }
+}
+
+function mergeChildNodes(
+  node: StoreLayer | null,
+  baseNode: BaseStoreLayer | NetworkResponseStoreLayer,
+): OptimisticStoreLayer | null {
+  while (node && node.kind !== 'OptimisticStoreLayer') {
+    mergeDataLayer(baseNode.data, node.data);
+    node = node.childStoreLayer;
+  }
+  return node;
+}
+
+function reexecuteUpdates(
+  node:
+    | OptimisticStoreLayer
+    | NetworkResponseStoreLayer
+    | StartUpdateStoreLayer
+    | null,
+  oldData: StoreLayerData,
+  newData: StoreLayerData,
+): void {
+  while (node !== null) {
+    mergeDataLayer(oldData, node.data);
+    if ('startUpdate' in node) {
+      node.data = {};
+      node.startUpdate(node);
+    }
+    mergeDataLayer(newData, node.data);
+
+    node = node.childStoreLayer;
+  }
+}
+
+function replaceOptimisticStoreLayerWithNetworkResponseStoreLayer(
+  environment: IsographEnvironment,
+  optimisticNode: OptimisticStoreLayer,
+  normalizeData: (storeLayer: StoreLayer) => void,
+): void {
+  const oldData = optimisticNode.data;
+  //  we cannot replace optimistic node with network response directly
+  // because of the types so we have to:
+  //  1. reset the optimistic node
+  optimisticNode.data = {};
+  // 2. append the network response as child
```

Comment:
Can we:

- store the optimistic layer's parent layer P1 and child layer C1
- merge the network response into the parent, getting a new parent P2 (which has type `BaseStoreLayer | NetworkResponseStoreLayer`)
- set the new parent P2's child to be C1... But this won't typecheck, so we need to actually do `p2.child = mergeChildNodes(c1, p2)`
- re-execute updates, but starting from `p2.child`.

That seems a bit simpler than what is going on here. I haven't really read this in a huge amount of detail yet, though. So maybe we're doing the same thing.